ad_lio:

  # Topics
  pointCloudTopic: "/ruby128"               # Point cloud data
  imuTopic: "gpsimu"                        # IMU data
  odomTopic: "odometry/imu"                 # IMU pre-preintegration odometry, same frequency as IMU
  gpsTopic: "odometry/gpsimu"               # GPS odometry topic from navsat, see module_navsat.launch file

  # Frames
  lidarFrame: "base_link"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  # GPS Settings
  useImuHeadingInitialization: true           # if using GPS data, set to "true"
  useGpsElevation: false                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data
  useGPSFactor: true                          # false true
  useGPSTimeGap: 0.01                         # s 0.01
  
  # Export settings
  savePCD: false
  savePCDDirectory: "/Project2024/LIO-SAM-Modified2024/src/pcd"        
  # in your home folder, starts and ends with "/". Warning the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation
  saveTraDirectory: "/home/whq/Project2024/LIO-SAM-Modified2024/src/AD-kd-shot-add.txt"

  # Sensor Settings
  sensor: robosense                           # lidar sensor type, 'velodyne' or 'ouster' or 'livox' or 'robosense'
  N_SCAN: 128                                 # number of lidar channel (i.e., Velodyne/Ouster: 16, 32, 64, 128, Livox Horizon: 6)
  Horizon_SCAN: 1800                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048, Livox Horizon: 4000)
  downsampleRate: 1                            # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
  lidarMinRange: 1.6                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 1000.0                       # default: 1000.0, maximum lidar range to be used
  point_filter_num: 2                         # special 1
  
  # liDAR Settings
  has_ring: false                             # lidar data format, if lidar driver has ring data publish, set 'ture'
  ang_bottom: -25                             # if 'has_ring' is false, change it
  ang_res_y: 0.1                              # if 'has_ring' is false, change it

  # IMU Settings
  #imuRate: 10
  imuRate: 100
  imuAccNoise: 3.9939570888238808e-03
  imuGyrNoise: 1.5636343949698187e-03
  imuAccBiasN: 6.4356659353532566e-05
  imuGyrBiasN: 3.5640318696367613e-05
  imuGravity: 9.7947 #9.80511
  imuRPYWeight: 0.01

  # Extrinsics: T_lb (lidar -> imu)
  extrinsicTrans: [0.0, 0.0, 0.0]
  lidarHeight: 2.4
  extrinsicRot: [ 0, 1, 0,
                  -1, 0, 0,
                  0, 0, 1 ]
  extrinsicRPY: [ 1, 0, 0,
                  0, 1, 0,
                  0, 0, 1 ]


  # LOAM feature threshold
  edgeThreshold: 1.0
  surfThreshold: 0.6
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.5                     # default: 0.4 - outdoor, 0.2 - indoor  0.4  tunnel 0.4  shiyanshi 0.5
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor  0.4  tunnel 0.4  shiyanshi 0.4
  mappingSurfLeafSize: 0.5                      # default: 0.4 - outdoor, 0.2 - indoor  0.4  tunnel 0.4  shiyanshi 0.5  downSample:1-4 0.5 downSample:5-7 0.05  origin 0.5 speial 0.2

  # adaptive downsample paprams
  initialSurfLeafSize: 0.5       #downSample:1-4 0.5  downSample:5-7 0.2
  initialSurfPointsSize: 3500    #downSample:1-4 3500 downSample:5-7 5000
  minSurfLeafSize: 0.2           #downSample:1-4 0.2  downSample:5-7 0.1  origin 0.2 speial 0.1
  maxSurfLeafSize: 1.0           #downSample:1-4 1.0  downSample:5-7 0.5  origin 1.0
  degenerateBoundary: 0.5
  adaptiveRate: 0.8

  # robot motion constraint (in case you are using a 2D robot)
  use_tollerance: false
  z_tollerance: 0.5                           # meters
  rotation_tollerance: 0.087                  # radians
  pitch_rotation_tollerance: 0.05             # radians

  # CPU Params
  numberOfCores: 4                              # number of cores for mapping optimization
  mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0  # meters, regulate keyframe adding threshold  1.0 0.2
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   2.0 0.5
  surroundingKeyframeSearchRadius: 50         # meters, within n meters scan-to-map optimization (when loop closure disabled) 50.0

  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density

# Navsat (convert GPS coordinates to Cartesian)
navsat:
  frequency: 50
  wait_for_datum: false
  delay: 0.0
  magnetic_declination_radians: 0
  yaw_offset: 0
  zero_altitude: true
  broadcast_utm_transform: false
  broadcast_utm_transform_as_parent_frame: false
  publish_filtered_gps: false

# EKF for Navsat
ekf_gps:
  publish_tf: false
  map_frame: map
  odom_frame: odom
  base_link_frame: base_link
  world_frame: odom

  frequency: 50
  two_d_mode: false
  sensor_timeout: 0.01
  # -------------------------------------
  # External IMU:
  # -------------------------------------
  imu0: imu_correct
  # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
  imu0_config: [false, false, false,
                true,  true,  true,
                false, false, false,
                false, false, true,
                true,  true,  true]
  imu0_differential: false
  imu0_queue_size: 50 
  imu0_remove_gravitational_acceleration: true
  # -------------------------------------
  # Odometry (From Navsat):
  # -------------------------------------
  odom0: odometry/gps
  odom0_config: [true,  true,  true,
                 false, false, false,
                 false, false, false,
                 false, false, false,
                 false, false, false]
  odom0_differential: false
  odom0_queue_size: 10

  #                            x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
  process_noise_covariance: [  1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015]